### DISCUSSION TITLE
initQpStatefulProperties() setting

### DISCUSSION THREAD

**Discussion by Joseph-0123** [idaholab/moose#23263] [Q&A General] [answered]
Dear all, 

I have a question about the Stateful Material Properties. I need to define the Stateful Material Properties _n1 and _n2. 

See below ：

_n1 is defined as the  _ind[_qp] multiplying coupledValue _s[_qp]. _ind[_qp] is the function of temperature _T[_qp]. The initial value of _n2 is 0. 

Are the following initQpStatefulProperties correct or not?

```
void
Material::initQpStatefulProperties()
{
  if (_T[_qp] < 1) {
    _ind[_qp] = 1.0;
  }
  else {
    _ind[_qp] = 0.0;
  }

  _n1[_qp] = _ind[_qp] * _s[_qp];
  _n2[_qp] = 0.0;
}
```

J

**Comment by Joseph-0123** [comment, marked as answer]
Hello @GiudGiud, could you please help me have a look at this?

**Reply by GiudGiud** [reply]
This seems right, depending on the definition that _ind should have which you did not explain.

However it's not the whole picture, this is just the initialization. You need to have a `computeQpPoperties` still
More on stateful properties
https://mooseframework.inl.gov/syntax/Materials/#f5ffdc6c-3cc9-4a43-bbfd-566e34905cf3

**Reply by Joseph-0123** [reply]
> This seems right, depending on the definition that _ind should have which you did not explain.
> 
> However it's not the whole picture, this is just the initialization. You need to have a `computeQpPoperties` still More on stateful properties https://mooseframework.inl.gov/syntax/Materials/#f5ffdc6c-3cc9-4a43-bbfd-566e34905cf3

Thanks for your answers, GiudGiud. I have another simple question:

in the computeQpJacobian of  TimeDerivative kernel: could 'du_dot_du' be replaced by '_phi[_j][_qp] * 1/dt' since the du_dot_du = d[(u-u_old)/dt]/du = _phi[_j][_qp] * 1/dt

**Reply by GiudGiud** [reply]
not if you want to use other more complicated time integration schemes. 
In general you want to use the dot routines to get XX_dot terms, they make sure the time derivative expression is the one consistent with the time integration scheme

**Reply by Joseph-0123** [reply]
> not if you want to use other more complicated time integration schemes. In general you want to use the dot routines to get XX_dot terms, they make sure the time derivative expression is the one consistent with the time integration scheme

Thanks a lot.
